active/0000-extensible-commands.md

```diff
@@ -0,0 +1,62 @@
+
+This lookup process would be refactored to use the DAG structure of loaded addons to build a flattened list of addons that implement the requested command. This stack would then be recursively invoked, such that each addon's version of the command can call `this._super()` to pass control to the next addon in the "stack". The bottom of the stack would be the built-in implementation (if one exists).
+
+This structure mimics Connect-style middleware, where each function in the stack can decide to pass control further, or halt the recursion by not calling `this._super()`.
+
+Here's a rough sketch in code:
+
+```
+var command, commandOptions;
+var orderedAddonList = project.addons;
+var commandStack = orderedAddonList.filter(function(addon) {
+        return addon.hasCommand(command);
+    }).map(function(addon) {
+        return addon.getCommand(command);
+    });
+
+var context = { _super: invokeNextCommandInStack };
+function invokeNextCommandInStack() {
+    var nextCommand = commandStack.shift();
+    return nextCommand.call(context, command, commandOptions);
+}
+
+invokeNextCommandInStack()
+```
+
+> **Note:** obviously, there's a lot more that should happen, and the commands are not functions themselves. Likely, either the Command model would need to incorporate the idea of a "stack", the `/lib/cli/cli.js` file would have to become stack-aware, or we introduce a new CommandStack abstraction to handle this.
+
+# Drawbacks
+
+* Extending built-in commands could result in user confusion, i.e. I checkout an Ember CLI project and run `ember test` and it performs unexpected behaviors, so I Google around for "ember test" and none of the standard docs help me.
+* Extending built-in commands could be a bit of a footgun - it would allow the addon developer to deviate from the conventions that this project attempts to cultivate.
+* Using the implementation approach outlined above, the addon that fires first for a built-in command could forgot to call super, causing a break in the chain and the default functionality not to run.
+
+# Alternatives
+
+* Rather than allowing complete overrides of commands, we could just provide hooks for common operations (i.e. modify the `availableOptions` hash, `anonymousOptions` array, etc). This would help prevent abuse, but limits use cases.
+* Even more restrictive, we could not allow any addon modification, and instead wait for common patterns or use cases to arise, and bake those into the built-in commands via optional arguments, auto-detecting addons, config, etc. This is difficult over the long term though, and doesn't scale well.
+* Finally, we may not want to allow customization of built-in commands at all, if the potential for abuse and surprising behavior outweighs the benefits of customization.
```

Comment:
That's an interesting view of the commands, maybe it's a new ember-cli thing, but I've never viewed commands on the command line as abstractions. 

Depending on other's thoughts, I think we could definitely come up with a way to provide context if we decide to allow people to override commands. Currently we allow generators to override the defaults which is a similar case. When you run `ember g -h` it outputs it in the section for the addon that provided it, so we could do the same thing with `ember -h`. It could be pretty annoying if every time you ran the overridden command it bothered you about it being from an addon